test/integration/controller/job/job_controller_test.go

```diff
@@ -831,4 +831,65 @@ var _ = ginkgo.Describe("Job controller interacting with scheduler", func() {
 			return createdProdJob.Spec.Suspend
 		}, util.Timeout, util.Interval).Should(gomega.Equal(pointer.Bool(false)))
 	})
+
+	ginkgo.When("The workload is deleted while it's admitted", func() {
+		ginkgo.It("Should restore the original node selectors", func() {
+
+			localQueue := testing.MakeLocalQueue("local-queue", ns.Name).ClusterQueue(prodClusterQ.Name).Obj()
+			job := testingjob.MakeJob(jobName, ns.Name).Queue(localQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
+			lookupKey := types.NamespacedName{Name: job.Name, Namespace: job.Namespace}
+			createdJob := &batchv1.Job{}
+
+			ginkgo.By("create a job", func() {
+				gomega.Expect(k8sClient.Create(ctx, job)).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("job should be suspend", func() {
+				gomega.Eventually(func() *bool {
+					gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob)).Should(gomega.Succeed())
+					return createdJob.Spec.Suspend
+				}, util.Timeout, util.Interval).Should(gomega.Equal(pointer.Bool(true)))
+			})
+
+			// backup the the podSet's node selector
+			initialSelector := createdJob.Spec.Template.Spec.NodeSelector
+
+			ginkgo.By("create a localQueue", func() {
+				gomega.Expect(k8sClient.Create(ctx, localQueue)).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("job should be unsuspended", func() {
+				gomega.Eventually(func() *bool {
+					gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob)).Should(gomega.Succeed())
+					return createdJob.Spec.Suspend
+				}, util.Timeout, util.Interval).Should(gomega.Equal(pointer.Bool(false)))
+			})
+
+			ginkgo.By("the node selector should be updated", func() {
+				gomega.Eventually(func() map[string]string {
+					gomega.Expect(k8sClient.Get(ctx, lookupKey, createdJob)).Should(gomega.Succeed())
+					return createdJob.Spec.Template.Spec.NodeSelector
+				}, util.Timeout, util.Interval).ShouldNot(gomega.Equal(initialSelector))
+			})
+
+			ginkgo.By("delete the workload", func() {
+				wl := &kueue.Workload{}
+				wlKey := types.NamespacedName{Name: workloadjob.GetWorkloadNameForJob(job.Name), Namespace: job.Namespace}
+				gomega.Expect(k8sClient.Get(ctx, wlKey, wl)).Should(gomega.Succeed())
+				gomega.Expect(util.DeleteWorkload(ctx, k8sClient, wl)).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("delete the localQueue, prevent readmission", func() {
```

Comment:
IIUC it is possible that the workload is re-admitted between ""delete the workload" and "delete the localQueue, prevent readmission" steps. Thus, making the "delete the workload" step unnecessary. Would the test pass if you delete it?

> creating something with higher priority might stop the job while the workload is in canceling admission and not because the workload gets deleted. 

What I meant is to create another (second) job, before the "delete the workload" step. Then, when we delete the workload, and the first job is unadmitted, scheduler will pick the second job with the higher priority. The second job admitted would block admission of the first job because it would use all resources. This way we will be able to check the node selectors of the first (unadmitted due to deleted workload) job without a risk of race conditions.

Reply:
> Thus, making the "delete the workload" step unnecessary. Would the test pass if you delete it?

No it will not, the workload will not be deleted in one `util.Timeout` interval.

However , we just need the job to be stopped and stay in a suspended  state, and it works.

Reply:
There is a race condition though in the test. Just insert
```golang 
time.Sleep(time.Second)
```
between "delete the workload" and "delete the localQueue, prevent readmission". I assume this could be prevented by the second workload. 

But on the other hand, do we benefit from scheduler in this test? I assume we can simplify the test and move to the section without scheduler (see related comment).